Extensible BoatEntity.Type

cech12 opened the pull request:
This PR adds the possibility to extend the BoatEntity.Type Enum via IExtensibleEnum interface. Its the secound try of the PR #6229 and a new PR because of too many changes.

It enables the possibility to add new boat types. For example, when a mod adds a new wood type, it can now add the new look also to the existing boat entity.

BoatEntity.Type got a new constructor, where the drops can be defined. Via the setBoatItem method it is possible to set the corresponding boat item. It is not in constructor because a BoatItem needs a BoatEntity.Type in constructor. The BoatRenderer generates the resource location of the textures via the BoatEntity.Type name. So "modId:name" as name is possible and recommended. The texture must be in directory textures/entity/boat and must have the name of the boat type.

The added test mod adds a green test boat.

Use Case: I want to create a mod with a new wood type and I want to add a new texture for boat entity.

cech12 committed:
Merge pull request #2 from MinecraftForge/1.14.x

update

cech12 committed:
Merge pull request #3 from MinecraftForge/1.14.x

Update

cech12 committed:
Merge pull request #4 from MinecraftForge/1.14.x

Update

cech12 committed:
Merge pull request #5 from MinecraftForge/1.14.x

update

cech12 committed:
Add IExtensibleEnum to vanilla BoatEntity.Type
- delegate the method getItemBoat to Type
- delegate the generated drops to Type
- add a setBoatItem method to Type because of the type dependency in BoatItem
- boat renderer generates the resource location of the textures with the type name (modId:name is also possible and recommended)
- added Test mod shows an example

tterrag1098 commented on patches/minecraft/net/minecraft/entity/item/BoatEntity.java.patch:
```diff
@@ -30,3 +69,62 @@
++      }
++
++      public void setBoatItem(Item boatItem) {
++         this.boatItem = boatItem;
```
Mutable enums is not a good idea, and this is never used.

cech12 replied:
You are right. I realized the boat item now as a supplier. So the setter is removed.

tterrag1098 commented on patches/minecraft/net/minecraft/entity/item/BoatEntity.java.patch:
```diff
@@ -30,3 +69,62 @@
++      }
++
++      public Item getBoatItem() {
++         return (this.boatItem == null) ? OAK.boatItem : this.boatItem;
```
This impl is useless, as all constants override it. Make the field a supplier instead to avoid needing anonymous enum classes.

cech12 replied:
Thanks for your hint with the supplier. I realized the boat item as a supplier now. So the getter works different.

tterrag1098 commented on patches/minecraft/net/minecraft/entity/item/BoatEntity.java.patch:
```diff
@@ -30,3 +69,62 @@
++      }
++
++      public net.minecraft.util.IItemProvider[] getDrops() {
++         return this.drops;
```
This should return an immutable view or keep the items as an immutable collection. Also, might be more flexible to use itemstacks?

cech12 replied:
Immutable set and the usage of itemstacks are implemented now. Thanks for all your feedback!

cech12 committed:
implement review comments:
- made boat item variable to supplier to avoid mutable enums (setter removed, getter adapted)
- drops are also realized by suppliers
- drops are now itemstacks
- getDrops returns now an immutable set

cech12 commented:
I implemented the change request. What do you think?

LexManos commented on patches/minecraft/net/minecraft/entity/item/BoatEntity.java.patch:
```diff
@@ -1,6 +1,29 @@
+-      case DARK_OAK:
+-         return Items.field_185154_aL;
+-      }
++      return this.func_184453_r().getBoatItem();
```
This patch can be so much smaller by moving this ABOVE the switch and letting vanilla types use the switch.

LexManos commented on patches/minecraft/net/minecraft/entity/item/BoatEntity.java.patch:
```diff
@@ -9,7 +32,23 @@
+-                     for(int j = 0; j < 2; ++j) {
+-                        this.func_199703_a(Items.field_151055_y);
+-                     }
+                   }
```
This isn't *bad*, but it seems like a area where Mojang hasn't moved to loot tables yet. How hard would it be to do that on our end instead of hardcoding.

LexManos commented on patches/minecraft/net/minecraft/entity/item/BoatEntity.java.patch:
```diff
@@ -30,3 +69,61 @@
++         return this.boatItemSupplier.get();
++      }
++
++      public com.google.common.collect.ImmutableSet<net.minecraft.item.ItemStack> getDrops() {
```
Typically we use Lists not sets to not limit duplication, this also doesn't have the same behavior of vanilla. Namely it'll toss out two entities that may or may not merge. Instead of 5 entities in random directions.

LexManos commented on patches/minecraft/net/minecraft/entity/item/BoatEntity.java.patch:
```diff
@@ -30,3 +69,61 @@
++      }
++
++      public Item getBoatItem() {
++         return this.boatItemSupplier.get();
```
Boat item sounds.. odd. And with lack of comments it took me a second to understand what this is for.

LexManos commented on patches/minecraft/net/minecraft/entity/item/BoatEntity.java.patch:
```diff
@@ -30,3 +69,61 @@
+          this.field_195934_h = p_i48146_3_;
++         this.boatItemSupplier = boatItemSupplier;
++         this.dropSupplier = () -> new net.minecraft.item.ItemStack[]{ new net.minecraft.item.ItemStack(Items.field_151055_y, 2), new net.minecraft.item.ItemStack(p_i48146_3_, 3) };
+       }
```
We typically do method redirects to prevent duplicates/lower patch size.

LexManos reviewed (changes requested):
This has some issues as marked in my other comments but the main one is that it breaks out LTS rules.
As this is targeting 1.14.4 without a 1.15 version.
@ichttt should of noticed that first off before assigning it to me.
